Make Taint command use Resource Address

huydinhle opened the pull request:
### Issue 
  - https://github.com/hashicorp/terraform/issues/11045

### Doc Update
  - I don't think we need doc update on the website, but we can clarify it if we really need to
  - I think the release note should be enough for this issue

### Relevant Terraform version
  - This is not urgent to be deployed immediately
  - It can wait for the next release of terraform

### About Acceptance Test.
  - I'll be running acceptance test in aws 
  - **I'm still working on writing the acceptance test for this code**

### Questions
  - Do I need to run acceptance test on Azure and Google also?
  - I ran the acceptance on aws and found no problem, should I save the log and report it here?
@heraclmene @apparentlymart

josephholsten commented:
@apparentlymart seems like this is a breaking change, but wanted: https://github.com/hashicorp/terraform/issues/11045#issuecomment-271174294

apparentlymart commented:
Indeed this seems like a good change for consistency, and I think the breaking change shouldn't be _too_ bad since the resource address syntax is pretty similar to what we previously expected, some details notwithstanding.

We don't actually expect _acceptance_ tests for this sort of core functionality, in the same sense as we would for providers, but rather they are usually tested via a mixture of command-line-level tests in the `command` package and internal-ish tests in the `terraform/context_*_test.go` tests.

I think this change is definitely a good idea though I think it might be a little close to 0.10 to add it to the scope of that release now. I'll pencil it in to the list and we'll see if we have time to look at it in more detail before the release. If not, we'll look at it for a later release.

Thanks for working on this, @huydinhle!

huydinhle commented:
@apparentlymart no problem at all!

apparentlymart commented on command/taint.go:
```diff
@@ -43,18 +44,20 @@ func (c *TaintCommand) Run(args []string) int {
+		resourceAddressStr = "module.root." + name
 	} else {
-		module = "root." + module
+		resourceAddressStr = "module.root.module." + module + "." + name
```
For resource addresses we actually don't expect any prefix for the root module. That is, a resource `aws_instance.baz` in the root module just has the resource address `aws_instance.baz` rather than `module.root.aws_instance.baz`. The `module.blah.` prefix is only used to traverse through child modules.

I think we could actually drop support for the `-module` flag here, given that we're considering this a breaking change anyway. Tainting a resource in a child module can be expressed as `terraform taint module.foo.aws_instance.baz` once we're using the standard resource address syntax.

huydinhle replied:
@apparentlymart Can you help me on this since I haven't worked on this for a while? If I don't put use `module.root.aws_instance.baz`as resource address then when I call `terraform.ParseResourceAddress(resourceAddressStr)` always give me `rsa.Path = []`  in turns make the `state.ModuleByPath` retun nil ---> making taint can't find resource state

apparentlymart commented on command/taint.go:
```diff
@@ -105,43 +109,16 @@ func (c *TaintCommand) Run(args []string) int {
-			return c.allowMissingExit(name, module)
-		}
+	// Get resource type from name
+	resourceType := strings.Split(name, ".")[0]
```
If I'm understanding correctly, this information is available in `rsa.Type`, and so we shouldn't need to extract it and pass it around as a separate argument.

apparentlymart commented on terraform/state.go:
```diff
@@ -185,6 +185,41 @@ func (s *State) moduleByPath(path []string) *ModuleState {
+	// If there are no resources in this module, it is an error
+	if len(mod.Resources) == 0 {
+		return nil, fmt.Errorf(
+			"The module %s has no resources. There is nothing to taint.", module)
```
I assume this error message was from before you did some refactoring... with the current approach, it's not guaranteed that the caller here is retrieving the resource state in order to taint it.

Given that this function is general and in core, I'd suggest that rather than returning an error we instead return `nil` in all of these error cases and then produce suitable error messages in the UI code, where we have more context available about what operation is being performed.

huydinhle commented:
@apparentlymart I would love to be able to finish the PR unless we are in a rush to have the feature in.  My bad for getting back to you late.

apparentlymart commented:
Hi @huydinhle!

By all means you are welcome to finish this yourself! The only slight time pressure here is that the next 0.10 beta is likely to represent a feature freeze, so I'd ideally like to land this in the next week or so, to make sure there's enough runway to get it in there. However, I don't expect you to rush to meet our schedule! If all else fails, we can include it in a future release.

Thanks again!

huydinhle commented:
@apparentlymart I think I addressed all the comments. Can you let me know if I need to do anything else. I would try to be actively working on this PR to get it merged. Thank you !

apparentlymart commented:
Hi @huydinhle! Sorry for the long silence.

Unfortunately I didn't see your comment here until today :confounded: and so I missed reviewing this for the 0.11 release too.

I've labelled this with breaking-change to remind us to look at it again for merging as part of the forthcoming 0.12 release. Since this changes some behavior of `terraform taint` we need to wait for a major release to give users a chance to update any automation they may have built around this command that depends on the old usage.

Sorry again for not noticing your updates and comment here before. This is a change I'm excited to merge since indeed the current difference in usage of this command is very confusing.

sonnyvan commented:
Any updates on this? Just wanted to follow up as we recently ran into this issue on my team with the `terraform taint` resource addressing.

huydinhle commented:
@apparentlymart  it's no problem at all, I would love to continue to work on this PR to get it merged. Please let me know if anything else need to be done

huydinhle commented:
@sonnyvan  let's wait for the next steps from Terraform team :smile:

huydinhle commented:
@apparentlymart  I haven't touched base with terraform in a while, can you give me some help to finish this PR ? thank you

pietermees commented:
@apparentlymart is this making it into 0.12?

apparentlymart commented:
Hi all! Sorry for the long silence here.

The work for v0.12 includes some significant changes to how state is modeled internally, and so we will need to change the implementation of the `terraform taint` command anyway in order to update it for the new internal models. When we make those changes we'll also adapt the work from this PR so that it'll use the standard resource address syntax, and merge all of these changes together.

apparentlymart commented:
Hi again @huydinhle! Sorry for the long silence here.

The work for 0.12 in the end included writing a new parser for the resource address syntax that uses the same underlying mechanism as the configuration language itself. That means that resource addresses will always use the same syntax as references within the configuration.

Unfortunately as a result of this the `taint` command implementation changed a lot. While fixing that, we integrated the new parser and removed the `-module` option at the same time, which integrated the _effect_ of your PR but written in a different way. This is why the changes in this PR are now conflicting with `master`.

I'm sorry that the changes here led to an independent implementation of what you implemented, rather than an adaptation of what you wrote as I originally suggested. I'm very grateful for the time you spent on this PR, and the discussions we had about it informed the design of the new implementation.

Since the functionality of this PR is now in master, I'm going to close this out. Thanks again for working on this, and I'm sorry again that the v0.12 work came into conflict with what you'd implemented here.
